### ISSUE TITLE
TLSv1 SSL handshake doesn't work with gtalk

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3463]
|<img alt="sjoerd's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| sjoerd reported|
|-|-|
|Trac ID|trac#3463|
|Type|enhancement|
|Created|2008-10-08 14:46:22Z|
|Branch|https://github.com/twisted/twisted/tree/jabber-tls-handshake-3463|

When connecting to gtalk using starttls and doing a TLSv1 SSL handshake, the gtalk server will stop responding and drop your connection :(

Using the SSLv23 method instead works fine, both for gtalk and various other jabber servers.

Attachments:

* [twisted-ssl-handshake.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ff0/ff0cb26418950ebc233178323fdf9c1a67a7dfac/afd173b625a8ab54b8b709e0c3eb4c700c220747.patch) (748 bytes) - added by sjoerd on 2008-10-08 14:46:52Z - 
* [tls-no-ticket.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ff0/ff0cb26418950ebc233178323fdf9c1a67a7dfac/4ec7fbdf76fb0f3fb3efba255b0e5d66931a1072.patch) (539 bytes) - added by ralphm on 2008-11-18 10:52:53Z - Patch to disable the Session Ticket extension

<details><summary>Searchable metadata</summary>

```
trac-id__3463 3463
type__enhancement enhancement
reporter__sjoerd sjoerd
priority__normal normal
milestone__Twisted_8_2 Twisted-8.2
branch__branches_jabber_tls_handshake_3463 branches/jabber-tls-handshake-3463
branch_author__ralphm ralphm
status__closed closed
resolution__fixed fixed
component__words words
keywords__ 
time__1223477182000000 1223477182000000
changetime__1227205582000000 1227205582000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_1"></a>|@ralphm set owner to @ralphm|
|-|-|

[XMPP Core (RFC 3920)](http://xmpp.org/rfcs/rfc3920.html) uses TLS for securing the transport, and only references SSL once to refer to older methods that did not use STARTTLS. Therefore, I believe this bug is at Google Talk but I will verify this within the Jabber community.

That said, if I am not mistaking, SSLv2 has several nasty vulnerabilities, so we don't want to use that. It seems that `SSLv3_METHOD` also works with Google Talk. However, I don't think TLS is used in this case, which would be a bad default.

That leaves your suggestion of using `SSLv23_METHOD`. The client would send SSLv2 hello messages, and advertise supporting SSLv2, SSLv3 and TLSv1. Those can then be limited to only TLSv1 by using `set_options` on the context (`OP_NO_SSLv2`, `OP_NO_SSLv3`). However, that doesn't seem to work with Google Talk. Disabling TLSv1 support and leaving SSLv3 support on, **does** work. That would suggest that Google Talk only supports SSLv3, and only if TLS is disabled, or none are disabled. Yuck. I'll look further into this.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_2"></a>|@ralphm commented|
|-|-|

Oh, by the way, this is the first time I noticed this. It has worked before, I'm quite sure.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

Other people have mentioned this problem before, but I've also been able to successfully connect to gtalk previously.

**Comment by twisted-trac** [comment]
|<img alt="sjoerd's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|sjoerd commented|
|-|-|

Some other people here that hit this issue noticed that old-style ssl works fine. But the <starttls /> method doesn't

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_5"></a>|@ralphm commented|
|-|-|

As mentioned on the Twisted Jabber mailinglist, the problem appears to be not in Google Talk per se. Apparently Java does not handle the empty Session Ticket extension in the hello, that signals support for the extension, correctly.

A workaround is to disable this extension by setting the `OP_NO_TICKET` option on the SSL context. I've verified that this allows the Twisted based clients to connect again. I attached a patch for a quick fix, but we need to do this better. First of all, my PyOpenSSL doesn't expose the option value (`0x4000`) with a module attribute, so I had to hard code it.

exarkun: do you have a suggestion how to do this properly within Twisted?

**Comment by twisted-trac** [comment]
|<img alt="WizKid's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|WizKid commented|
|-|-|

I first reported the bug to Sun the 22 July.

When Sun make the bug public, info about it will be here: http://bugs.sun.com/bugdatabase/view_bug.do?bug_id=6728126 . But when that happen I have no idea. First Sun thought that there could be some security problem with the bug but when they understood that it couldn't they made it public for a short time but then they removed it again.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_7"></a>|@ralphm commented|
|-|-|

(In [25410]) Branching to 'jabber-tls-handshake-3463'

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_8"></a>|@ralphm removed owner|
|-|-|

This change disables the Session Tickets extension by default, which makes connections to Google Talk and other buggy implementations work again.

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @ralphm|
|-|-|

  1. Please open a ticket in the pyOpenSSL tracker for exposing OP_NO_TICKET
  1. How can I test this?  I tried doc/words/examples/xmpp_client.py but it apparently managed to start TLS with and without the changes in this branch.  I suppose the OpenSSL version is important here.  Do you know which version introduces the problem?  I have 0.9.8g-4ubuntu3.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_11"></a>|@ralphm removed owner|
|-|-|

Replying to [exarkun](#note_10):
>   1. Please open a ticket in the pyOpenSSL tracker for exposing OP_NO_TICKET

https://bugs.launchpad.net/pyopenssl/+bug/300220 Done].

>   2. How can I test this?  I tried doc/words/examples/xmpp_client.py but it apparently managed to start TLS with and without the changes in this branch.  I suppose the OpenSSL version is important here.  Do you know which version introduces the problem?  I have 0.9.8g-4ubuntu3.

The addition of Session Tickets has been introduced in one of the versions that ship with Ubuntu Intrepid (8.10). This is somewhere after the version you mentioned, which is for Hardy (8.4). I haven't been able to track down exactly which version, though. Looking at the upstream changelog, this functionality should only appear in version 0.9.9, but Ubuntu has applied a large patch against 0.9.8g for its packages.

Although the Twisted 8.2 milestone has been deleted (why not just closed?), it would be very nice if this could ship in that release.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to @ralphm|
|-|-|

Great.  Please merge.

Talk to radix about how to get this into 8.2.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_12"></a>|@ralphm set status to `closed`|
|-|-|

(In [25471]) Add an option to select if TLS Session Ticket extension is used.

Author: ralphm.
Reviewer: exarkun.
Fixes [#3463](https://github.com/twisted/twisted/issues/3463).

The option is off by default to work around faulty server implementations that
become unresponsive when sending an empty TLS Session Ticket extension in the
TLS hello. This should solve connection issues with Google Talk, if using
an OpenSSL implementation that enables session tickets by default.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
